File: py/src/iconomics/coa.py
"""Chart of accounts: account code -> statement line.

Treated as data, not code. The shipped codes are illustrative and meant to be
replaced with the firm's actual сметкоплан — see the warning in config/coa.yaml.
"""

from dataclasses import dataclass
from pathlib import Path

import yaml

from iconomics.config import ConfigError, find_config_dir

PROFIT_AND_LOSS = "profit_and_loss"
BALANCE_SHEET = "balance_sheet"


class CoaError(ConfigError):
    """Raised when config/coa.yaml is missing or malformed."""


@dataclass(frozen=True)
class Account:
    code: str
    name: str
    statement: str
    line: str
    side: str
    cash: bool = False


@dataclass(frozen=True)
class Coa:
    accounts: dict[str, Account]
    line_order: dict[str, list[str]]

    def get(self, code: str) -> Account | None:
        """Resolve a code, falling back to its parent for sub-accounts.

        A firm that books to 6021 rather than 602 should still roll up correctly,
        so lookup walks the code down one character at a time.
        """
        key = (code or "").strip()
        while key:
            if key in self.accounts:
                return self.accounts[key]
            key = key[:-1]
        return None

    def ordered_lines(self, statement: str) -> list[str]:
        return list(self.line_order.get(statement, []))

    def cash_accounts(self) -> list[str]:
        return sorted(code for code, a in self.accounts.items() if a.cash)


def load_coa(config_dir: Path | None = None) -> Coa:
    directory = config_dir if config_dir is not None else find_config_dir()
    path = Path(directory) / "coa.yaml"
    if not path.is_file():
        raise CoaError(f"missing config file: {path}")

    raw = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
    entries = raw.get("accounts") or {}
    if not entries:
        raise CoaError(f"{path} defines no accounts")

    accounts: dict[str, Account] = {}
    for code, spec in entries.items():
        statement = str(spec.get("statement", "")).strip()
        if statement not in (PROFIT_AND_LOSS, BALANCE_SHEET):
            raise CoaError(
                f"{path}: account {code} has statement {statement!r}; "
                f"expected {PROFIT_AND_LOSS} or {BALANCE_SHEET}"
            )
        side = str(spec.get("side", "")).strip()
        if side not in ("debit", "credit"):
            raise CoaError(
                f"{path}: account {code} has side {side!r}; expected debit or credit"
            )
        accounts[str(code).strip()] = Account(
            code=str(code).strip(),
            name=str(spec.get("name", "")).strip(),
            statement=statement,
            line=str(spec.get("line", "")).strip(),
            side=side,
            cash=bool(spec.get("cash", False)),
        )

    order = raw.get("line_order") or {}
    line_order = {
        key: [str(item) for item in (order.get(key) or [])]
        for key in (PROFIT_AND_LOSS, BALANCE_SHEET)
    }
    return Coa(accounts=accounts, line_order=line_order)

File: js/src/coa.js
/**
 * Chart of accounts: account code -> statement line.
 * Mirrors py/src/iconomics/coa.py.
 *
 * Treated as data, not code. The shipped codes are illustrative and meant to be
 * replaced with the firm's actual сметкоплан — see the warning in config/coa.yaml.
 */

import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import yaml from 'js-yaml';

import { ConfigError, findConfigDir } from './config.js';

export const PROFIT_AND_LOSS = 'profit_and_loss';
export const BALANCE_SHEET = 'balance_sheet';

export class CoaError extends ConfigError {}

export function loadCoa(configDir = null) {
  const directory = configDir ?? findConfigDir();
  const path = join(directory, 'coa.yaml');
  if (!existsSync(path)) throw new CoaError(`missing config file: ${path}`);

  const raw = yaml.load(readFileSync(path, 'utf-8')) ?? {};
  const entries = raw.accounts ?? {};
  if (Object.keys(entries).length === 0) throw new CoaError(`${path} defines no accounts`);

  const accounts = {};
  for (const [code, spec] of Object.entries(entries)) {
    const statement = String(spec.statement ?? '').trim();
    if (statement !== PROFIT_AND_LOSS && statement !== BALANCE_SHEET) {
      throw new CoaError(
        `${path}: account ${code} has statement '${statement}'; ` +
          `expected ${PROFIT_AND_LOSS} or ${BALANCE_SHEET}`,
      );
    }
    const side = String(spec.side ?? '').trim();
    if (side !== 'debit' && side !== 'credit') {
      throw new CoaError(
        `${path}: account ${code} has side '${side}'; expected debit or credit`,
      );
    }
    const key = String(code).trim();
    accounts[key] = {
      code: key,
      name: String(spec.name ?? '').trim(),
      statement,
      line: String(spec.line ?? '').trim(),
      side,
      cash: Boolean(spec.cash),
    };
  }

  const order = raw.line_order ?? {};
  const lineOrder = {
    [PROFIT_AND_LOSS]: (order[PROFIT_AND_LOSS] ?? []).map(String),
    [BALANCE_SHEET]: (order[BALANCE_SHEET] ?? []).map(String),
  };

  return {
    accounts,
    lineOrder,

    /**
     * Resolve a code, falling back to its parent for sub-accounts, so a firm
     * booking to 6021 still rolls up under 602.
     */
    get(code) {
      let key = String(code ?? '').trim();
      while (key) {
        if (key in accounts) return accounts[key];
        key = key.slice(0, -1);
      }
      return null;
    },

    orderedLines(statement) {
      return [...(lineOrder[statement] ?? [])];
    },

    cashAccounts() {
      return Object.keys(accounts)
        .filter((code) => accounts[code].cash)
        .sort();
    },
  };
}

/**
 * The normal side of the accounts mapped to this statement line.
 *
 * Resolved from the lexicographically smallest account code, not iteration
 * order: JavaScript objects reorder integer-like keys numerically while Python
 * dicts keep YAML order, so order-dependent logic would diverge.
 */
export function lineSide(lineName, coa) {
  for (const code of Object.keys(coa.accounts).sort()) {
    if (coa.accounts[code].line === lineName) return coa.accounts[code].side;
  }
  return 'debit';
}
